pageserver/src/tenant/secondary/downloader.rs

```diff
@@ -477,14 +486,58 @@ impl<'a> TenantDownloader<'a> {
         };
 
         let tenant_shard_id = self.secondary_state.get_tenant_shard_id();
+
+        // We will use the etag from last successful download to make the download conditional on changes
+        let last_etag = self
+            .secondary_state
+            .detail
+            .lock()
+            .unwrap()
+            .last_etag
+            .clone();
+
         // Download the tenant's heatmap
-        let heatmap_bytes = tokio::select!(
-            bytes = self.download_heatmap() => {bytes?},
+        let (heatmap_etag, heatmap_mtime, heatmap_bytes) = match tokio::select!(
+            bytes = self.download_heatmap(last_etag.as_ref()) => {bytes?},
             _ = self.secondary_state.cancel.cancelled() => return Ok(())
-        );
+        ) {
+            HeatmapDownload::Unmodified => {
+                tracing::info!("Heatmap unchanged since last successful download");
+                return Ok(());
+            }
+            HeatmapDownload::Modified(heatmap_etag, heatmap_mtime, heatmap_bytes) => {
+                (heatmap_etag, heatmap_mtime, heatmap_bytes)
+            }
+        };
 
         let heatmap = serde_json::from_slice::<HeatMapTenant>(&heatmap_bytes)?;
 
+        // {
+        //     let mut ondisk_bytes = 0;
+        //     let mut ondisk_layers = 0;
+        //     {
+        //         let detail = self.secondary_state.detail.lock().unwrap();
+        //         for (timeline_id, timeline) in detail.timelines.values() {
+        //             for (layer_name, layer) in timeline.on_disk_layers.values() {
+        //                 if let Some(hm_tl) = heatmap.timelines.get(timeline_id) {
+        //                     if hm_tl.l
+
+        //                     ondisk_layers += 1;
+        //                     ondisk_bytes += layer.metadata.file_size();
+        //                 }
+        //             }
+        //             timeline.on_disk_layers.len()
+        //         }
+        //     }
+        //     let mut progress = self.secondary_state.progress.lock().unwrap();
+        //     let heatmap_stats = heatmap.get_stats();
+        //     progress.layers_total = heatmap_stats.layers;
+        //     progress.bytes_total = heatmap_stats.bytes;
+        //     progress.heatmap_mtime = Some(heatmap_mtime);
+        //     progress.bytes_downloaded = ondisk_bytes;
+        //     progress.layers_downloaded = ondisk_layers;
+        // }
```

Comment:
Did you mean to remove this?

Reply:
Yes - 9b1b91359

Sorry, there are a few too many nits in this PR, I should have looked it over more carefully before opening for review.